src/gui/item_color_picker_2d.cpp

```diff
@@ -0,0 +1,143 @@
+//  SuperTux
+//  Copyright (C) 2024 HybridDog
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "gui/item_color_picker_2d.hpp"
+
+#include <vector>
+
+#include "math/util.hpp"
+#include "video/drawing_context.hpp"
+#include "video/video_system.hpp"
+#include "video/surface.hpp"
+#include "video/sdl_surface.hpp"
+
+
+namespace {
+
+/** Get the color of a pixel from an 8-bit RGB image
+ *
+ * @param surface SDL2 surface containing the image
+ * @param pos Floating-point pixel coordinates in [0,1]^2
+ *
+ * @return Color of the pixel at the integer position which corresponds to pos
+ */
+Color
+get_pixel(const SDLSurfacePtr& surface, const Vector& pos)
+{
+  assert(surface->format->BytesPerPixel == 3);
+  int x = math::clamp(
+    static_cast<int>(pos.x * static_cast<float>(surface->w - 1) + 0.5f),
+    0,
+    surface->w - 1
+  );
+  int y = math::clamp(
+    static_cast<int>(pos.y * static_cast<float>(surface->h - 1) + 0.5f),
+    0,
+    surface->h - 1
+  );
+  uint8_t *pixel = static_cast<uint8_t *>(surface->pixels)
+    + y * surface->pitch + x * 3;
+  if (SDL_BYTEORDER == SDL_BIG_ENDIAN)
+    return Color::from_rgb888(pixel[2], pixel[1], pixel[0]);
+  return Color::from_rgb888(pixel[0], pixel[1], pixel[2]);
```

Comment:
This can be replaced with preprocessor macro.

Reply:
I have replaced it with a constexpr condition, which I think is a bit cleaner than a macro.